### PR TITLE
CORTX-29500:  Error response body handled

### PR DESCRIPTION
**Pull request by abhijit1patil**
Signed-off-by: abhijit1patil <abhijit.1.patil@seagate.com>

# Problem Statement
- Bug 1 [CORTX-29506 HttpException Error Response](https://jts.seagate.com/browse/CORTX-29506)
- Bug 2 [CORTX-29500 CSM error Response](https://jts.seagate.com/browse/CORTX-29500)
- Bug 3 [CORTX-29454 Unsupported API ](https://jts.seagate.com/browse/CORTX-29454)
- Bug 4 [CORTX-29491 Inconsistent message](https://jts.seagate.com/browse/CORTX-29491)

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Build and Deployment
-  [Jenkins Custom Build Job](https://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/generic/job/custom-ci/5605/)
- Custom Builds:-
cortx-docker.colo.seagate.com/seagate/cortx-all:2.0.0-5605-custom-ci
cortx-docker.colo.seagate.com/seagate/cortx-rgw:2.0.0-5605-custom-ci
cortx-docker.colo.seagate.com/seagate/cortx-data:2.0.0-5605-custom-ci

- [Jenkins Cluster Deployment Job](https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/RGW/job/setup-cortx-rgw-cluster/1965/)
- VM Cluster deployed
image

# Dev Testing

- **For Web Exceptions - HTTP Error**  [Method Not Allowed]
image
- **For Web Exceptions - HTTP Error**  [Unsupported Endpoint]
image

- For CSM handled HTTP Error (Forbidden) 
1.  Login with Monitor CSM User
image
2.  Create a User with Monitor Bearer Token
image

- CSM Handled Http Error [Unauthorized Error] 
image

**For CSM Error**
- Testcase 3:- For CSM Error - Generated by CSM 
image

- Testcase 4:- For CSM Error - Generated by RGW
image

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: csm/common/errors.py [hunk 1]
```diff
@@ -48,8 +58,9 @@ class CsmSetupError(CsmError):
     """
     _desc = "Csm Setup is failed"
```
Comment (pawankumarsrivastava) [inline csm/common/errors.py:59]: Csm Setup is failed -> CSM setup failed.

File: csm/common/errors.py [hunk 2]
```diff
@@ -74,7 +85,7 @@ class InvalidRequest(CsmError):
     _err = CSM_INVALID_REQUEST
     _desc = "Invalid request message received."
```
Comment (pawankumarsrivastava) [inline csm/common/errors.py:86]: Invalid request message received. --> Invalid request.

File: csm/common/errors.py [hunk 3]
```diff
@@ -119,31 +134,55 @@ class CsmPermissionDenied(CsmError):
...
-            CSM_INTERNAL_ERROR, desc,
+            CSM_OPERATION_NOT_PERMITTED, desc,
             message_id, message_args)
 
 
 class CsmResourceNotAvailable(CsmInternalError):
 
     """Describes issues when requested resource is not available"""
 
+    _desc = "Requested resource is not available"
```
Comment (pawankumarsrivastava) [inline csm/common/errors.py:149]: Requested resource is not available --> Resource not available

File: csm/common/errors.py [hunk 4]
```diff
@@ -119,31 +134,55 @@ class CsmPermissionDenied(CsmError):
...
+
+    def __init__(self, desc=None, message_id=const.RESOURCE_NOT_AVAILABLE, message_args=None):
+        super(CsmResourceNotAvailable, self).__init__(
+            desc, message_id, message_args)
 
 class CsmTypeError(CsmInternalError):
 
     """Issues related to incorrect type of argument/parameter, etc."""
 
+    _desc = "Incorrect type of argument/parameter"
```
Comment (pawankumarsrivastava) [inline csm/common/errors.py:159]: Incorrect type of argument/parameter --> Unsupported type

File: csm/common/errors.py [hunk 5]
```diff
@@ -119,31 +134,55 @@ class CsmPermissionDenied(CsmError):
...
+
+    def __init__(self, desc=None, message_id=const.TYPE_ERROR, message_args=None):
+        super(CsmTypeError, self).__init__(
+            desc, message_id, message_args)
 
 class CsmNotImplemented(CsmError):
 
     """This error represents HTTP 501 Not Implemented Error"""
 
+    _desc = "Not Implemented Error"
```
Comment (pawankumarsrivastava) [inline csm/common/errors.py:169]: Not Implemented Error --> Not Implemented

File: csm/common/errors.py [hunk 6]
```diff
@@ -119,31 +134,55 @@ class CsmPermissionDenied(CsmError):
...
+    def __init__(self, desc=None, message_id=const.NOT_IMPLEMENTED, message_args=None):
+        super(CsmNotImplemented, self).__init__(
+            CSM_NOT_IMPLEMENTED, desc,
+            message_id, message_args)
 
 class CsmServiceConflict(CsmError):
 
     """Service in conflict stat or operation can cause that state"""
 
+    _desc = "Service conflict state error."
```
Comment (pawankumarsrivastava) [inline csm/common/errors.py:180]: Service conflict state error. --> Service conflict state.

File: csm/common/errors.py [hunk 7]
```diff
@@ -152,11 +191,42 @@ class CsmGatewayTimeout(CsmError):
     a timely response from the upstream server.
     """
 
+    _desc = "Did not receive a timely response from the upstream server"
```
Comment (pawankumarsrivastava) [inline csm/common/errors.py:194]: Did not receive a timely response from the upstream server --> Unable to get timely response

File: csm/common/errors.py [hunk 8]
```diff
@@ -152,11 +191,42 @@ class CsmGatewayTimeout(CsmError):
...
+    def __init__(self, desc=None, message_id=const.UNAUTHORIZED_ERROR, message_args=None):
+        super(CsmUnauthorizedError, self).__init__(
+            CSM_UNAUTHORIZED_ERROR, desc,
+            message_id, message_args)
+
 class CsmServiceNotAvailable(CsmError):
 
     """This  error represents CSM service is Not Available."""
+
+    _desc = "CSM service is Not Available."
```
Comment (pawankumarsrivastava) [inline csm/common/errors.py:216]: CSM service is Not Available. ---> CSM service not available

File: csm/common/errors.py [hunk 9]
```diff
@@ -152,11 +191,42 @@ class CsmGatewayTimeout(CsmError):
...
+    def __init__(self, desc=None, message_id=const.SERVICE_NOT_AVAILABLE, message_args=None):
+        super(CsmServiceNotAvailable, self).__init__(
+            CSM_SERVICE_NOT_AVAILABLE, desc,
+            message_id, message_args)
+
+class CsmRequestCancelled(CsmError):
+
+    """This  error represents CSM service request is cancelled."""
+
+    _desc = "CSM service request is cancelled."
```
Comment (pawankumarsrivastava) [inline csm/common/errors.py:227]: CSM service request is cancelled. --> Service request cancelled.

File: csm/core/controllers/routes.py [hunk 10]
```diff
@@ -20,7 +20,7 @@
 from .onboarding import OnboardingStateView
 from .system_config import SystemConfigListView
 from .system_config import SystemConfigView
-from .storage_capacity import StorageCapacityView
+# from .storage_capacity import StorageCapacityView
```
Comment (pawankumarsrivastava) [inline csm/core/controllers/routes.py:23]: If not required please remove this line.

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]

File: csm/common/errors.py (hunk 1 above)
Comment (abhijit1patil) [reply csm/common/errors.py]: updated

File: csm/common/errors.py (hunk 2 above)
Comment (abhijit1patil) [reply csm/common/errors.py]: updated

File: csm/common/errors.py (hunk 3 above)
Comment (abhijit1patil) [reply csm/common/errors.py]: For this error, message and message id will be similar 
message = "Resource not available"
message_id = "ResourceNotAvailable"
is it ok?

File: csm/common/errors.py (hunk 4 above)
Comment (abhijit1patil) [reply csm/common/errors.py]: updated

File: csm/common/errors.py (hunk 5 above)
Comment (abhijit1patil) [reply csm/common/errors.py]: For this error, message and message id will be similar
message = "Not Implemented"
message_id = "NotImplemented"
is it ok?

File: csm/common/errors.py (hunk 6 above)
Comment (abhijit1patil) [reply csm/common/errors.py]: updated

File: csm/common/errors.py (hunk 7 above)
Comment (abhijit1patil) [reply csm/common/errors.py]: updated

File: csm/common/errors.py (hunk 8 above)
Comment (abhijit1patil) [reply csm/common/errors.py]: updated

File: csm/common/errors.py (hunk 9 above)
Comment (abhijit1patil) [reply csm/common/errors.py]: updated

File: csm/core/controllers/routes.py (hunk 10 above)
Comment (abhijit1patil) [reply csm/core/controllers/routes.py]: updated

File: csm/core/agent/api.py [hunk 11]
```diff
@@ -169,17 +172,19 @@ def error_response(err: Exception, **kwargs) -> dict:
...
+            resp["message_id"] = err.message_id()
             message_args = err.message_args()
             if message_args is not None:
                 resp["error_format_args"] = err.message_args()
         elif isinstance(err, web_exceptions.HTTPError):
-            resp["message"] = str(err)
-            resp["error_code"] = err.status
+            resp["error_code"] = CSM_HTTP_ERROR
+            resp["message_id"] = str(err.reason)
+            resp["message"] = str(err.text)
```
Comment (rohitkolapkar) [inline csm/core/agent/api.py:182]: Please update,
resp["message_id"] = 'HTTPError'

add 'HTTPError' to const file and read this from const.

**Review by rohitkolapkar — CHANGES_REQUESTED** [review]

File: csm/core/agent/api.py (hunk 11 above)
Comment (abhijit1patil) [reply csm/core/agent/api.py]: updated

**Review by rohitkolapkar — APPROVED** [review]

**Review by pawankumarsrivastava — APPROVED** [review]

**Pull request merged**